### PR TITLE
Add ops_coverage python script

### PR DESCRIPTION
**Pull request by bhack**
@ricmk Do you want to support me to integrate this script in the CI?

Please check https://github.com/tensorflow/tensorflow/issues/14798#issuecomment-1047796247

/cc @andrew-leaver @cheshire

### CODE REVIEW

**Comment by bhack** [comment]
Check also my "extra" comments:
https://github.com/tensorflow/tensorflow/commit/8fccf8a7728315d1cf2c538a3bb55029103addac#commitcomment-75913049
https://github.com/tensorflow/tensorflow/commit/a099818c97085a7174da52dc05cf5b49ec7105cf#r76511015

File: tensorflow/compiler/tf2xla/g3doc/generate_ops_coverage.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+)
+
+writer_ops_list = MarkdownTableWriter(
+    table_name=table_name,
+    headers=fieldnames,
+    value_matrix=value_matrix,
+    margin=1
+)
+
+with open("tf_os_bridges_coverage.md", "w") as f:
```
Comment (cheshire) [inline tensorflow/compiler/tf2xla/g3doc/generate_ops_coverage.py:59]: Let's write to stdout by default, users can redirect if they need.

File: tensorflow/compiler/tf2xla/g3doc/generate_ops_coverage.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  if op.supports_new_bridge:
+    new_bridge_value = ":heavy_check_mark:"
+    new_bridge += 1
+  if (op.supports_old_bridge and op.supports_new_bridge):
+    all_bridges += 1
+  value_matrix.append([op.graph_op_name, old_bridge_value, new_bridge_value])
+
+writer = MarkdownTableWriter()
+
+writer_ops_stats = MarkdownTableWriter(
```
Comment (cheshire) [inline tensorflow/compiler/tf2xla/g3doc/generate_ops_coverage.py:39]: Bringing in a new dependency just to format a table does not quite seem worth it, maybe just do a helper function?

File: tensorflow/compiler/tf2xla/g3doc/generate_ops_coverage.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+
+today = date.today()
+today = today.strftime("%Y-%m-%d")
+
+supported_ops = Support.OpSupports()
+
+
+def parse_supported_ops():
+  with open("./support.pbtxt", 'r') as f:
```
Comment (cheshire) [inline tensorflow/compiler/tf2xla/g3doc/generate_ops_coverage.py:14]: Where does this file come from?

**Review by cheshire — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/tf2xla/g3doc/generate_ops_coverage.py (hunk 1 above)
Comment (bhack) [reply tensorflow/compiler/tf2xla/g3doc/generate_ops_coverage.py:59]: Honestly the scope of this was less for an user self-use but instead to be integrated in the build so when @ricmk makes  commits like this https://github.com/tensorflow/tensorflow/commit/a099818c97085a7174da52dc05cf5b49ec7105cf it is going to updated the markdown file.

File: tensorflow/compiler/tf2xla/g3doc/generate_ops_coverage.py (hunk 3 above)
Comment (bhack) [reply tensorflow/compiler/tf2xla/g3doc/generate_ops_coverage.py:14]: It is maintained by @rick at https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2xla/support.pbtxt but we need to compile the protobuf in the build system for importing it in  python (line 4).

**Comment by rick** [comment]
the-song

File: tensorflow/compiler/tf2xla/g3doc/generate_ops_coverage.py (hunk 3 above)
Comment (bhack) [reply tensorflow/compiler/tf2xla/g3doc/generate_ops_coverage.py:14]: @ricmk Can you join in this thread?

Comment (ricmk) [reply tensorflow/compiler/tf2xla/g3doc/generate_ops_coverage.py:14]: Hi @bhack, the support.pbtxt information has not been fully verified and is initially intended for deriving the XLA compilation support per TF op (and is not likely suitable for general coverage tables at the moment).

Comment (bhack) [reply tensorflow/compiler/tf2xla/g3doc/generate_ops_coverage.py:14]: > has not been fully verified and is initially intended for deriving the XLA compilation support per TF op

Is it manually populated? Or are you going to automate it using the [XLA op registry](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2xla/xla_op_registry.h) and other registers?

Also if your work is still WIP I still think that it is better then nothing that we produce a table as the official tables in the same folder are outdated by 4 years and I think it is enough time for a project like TF. 

So having a WIP table It is is better then just let the user fail the function compilation op after op (as you know we have a fail fast policy in `jit_compile`).

Comment (ricmk) [reply tensorflow/compiler/tf2xla/g3doc/generate_ops_coverage.py:14]: Hi bhack@, this protobuf design is not yet stable and is subject to changes (these fields in particular will be removed shortly), so we do not want to introduce any dependencies on its field at this time. Thank you for your understanding.

Comment (bhack) [reply tensorflow/compiler/tf2xla/g3doc/generate_ops_coverage.py:14]: >  so we do not want to introduce any dependencies on its field at this time

We could modify the script as you change the proto it is just something that it is going to produce a markdown page where now we have something that is not updated by more then 4 years. I suppose that we don't need to find the perfection of the definitive proto when we have something so outdated.

Also you have not replied to my second point:
> Is it manually populated? Or are you going to automate it using the [XLA op registry](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2xla/xla_op_registry.h) and other registers?

Comment (ricmk) [reply tensorflow/compiler/tf2xla/g3doc/generate_ops_coverage.py:14]: The protobuf is currently populated by a combination of static analysis and manual verification. We are relocating the protobuf shortly and will resurface the compilation data with more explicit meanings for the fields. Cheers.

**Comment by bhack** [comment]
@cheshire @ricmk Can you share a little bit the plan/roadmap after you have removed all the support protos with this commit
https://github.com/tensorflow/tensorflow/commit/38582d06e9acd5108ce943af1e793aa493282cad ?

**Comment by bhack** [comment]
@paynecl Any news on this?

File: tensorflow/compiler/tf2xla/g3doc/generate_ops_coverage.py (hunk 2 above)
Comment (MarkDaoust) [reply tensorflow/compiler/tf2xla/g3doc/generate_ops_coverage.py:39]: Maybe there's something in https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/docs

But I think those all generate html tables.

Comment (bhack) [reply tensorflow/compiler/tf2xla/g3doc/generate_ops_coverage.py:39]: The problem now is also where we are going to take the source of truth? As the protos were removed in the meantime.

**Comment by bhack** [comment]
Can we use/rely on `XlaOpRegistry` to generate this table? 
https://github.com/tensorflow/tensorflow/blob/814e437df95b8b1455452feb47eee5c5e37ce229/tensorflow/compiler/tf2xla/xla_op_registry.cc#L57

If this could be ok Is this only valid for the "standard" TF brdige?

**Comment by paynecl** [comment]
Yes.  I actually have an internal change for this that I'll send for review
that integrates the information with the raw ops table (sorry for the long
delay).
Current approach *would* use the opset defined by the registry of the non
milr bridge.  That is currently the default bridge for GPU/CPU so we'd use
that as ground truth in the docs. We'd switch over to the MLIR opset
construction when that bridge becomes default.

**Comment by bhack** [comment]
Thanks for the update, as it seems an internal change I suppose that it will be pushed directly on GitHub without a PR.

I will close this PR when you are ready as I cannot interact with a CL right now.

**Comment by bhack** [comment]
I don't know. @paynecl What is the status of your mentioned CL?

**Comment by MarkDaoust** [comment]
I think @paynecl is looking into it.

**Comment by paynecl** [comment]
Sorry for the delay.  Draft is written and I've synced with Mark on the
documentation updates; I need to get someone to help with an internal issue
but can submit after it is resolved.

**Comment by gbaned** [comment]
Hi @paynecl Any update on this CL? Please. Thank you!

**Comment by MarkDaoust** [comment]
@paynecl submitted the changes to the doc generator, I'm not 100% sure why this hasn't showed up on the raw_ops page yet:

https://www.tensorflow.org/api_docs/python/tf/raw_ops

**Comment by gbaned** [comment]
Hi @paynecl Any update on this PR? Please. Thank you!

**Comment by MarkDaoust** [comment]
Fixed: https://www.tensorflow.org/api_docs/python/tf/raw_ops

**Comment by bhack** [comment]
@MarkDaoust thanks. It would be really nice to know what is planned to be covered and what it will be never covered by design. I think it will help a lot to choice on what ops to rely if you need to compile your whole graph. 
Do you think is it possible to add this info in the page?

**Comment by gbaned** [comment]
Hi @MarkDaoust Any update on above [comment](https://github.com/tensorflow/tensorflow/pull/56510#issuecomment-1738999283) from @bhack? Please. Thank you!

**Comment by MarkDaoust** [comment]
> Do you think is it possible to add this info in the page?

No that information is not available.

**Pull request closed without merging**